File: android/app/src/main/kotlin/com/example/elm327emu/Elm327Plugin.kt
package com.example.elm327emu

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel

class Elm327Plugin(private val context: Context) {
    private var events: EventChannel.EventSink? = null
    private val main = Handler(Looper.getMainLooper())
    private var ble: BleGattServer? = null
    private var spp: SppServer? = null

    fun register(control: MethodChannel, eventChannel: EventChannel) {
        control.setMethodCallHandler { call, result ->
            when (call.method) {
                "capabilities" -> result.success(listOf("ble", "spp"))
                "startBle" -> {
                    val fff0 = call.argument<String>("profile") == "fff0"
                    ble?.stop()
                    ble = BleGattServer(context, rx("ble"), conn("ble")).also { it.start(fff0) }
                    result.success(null)
                }
                "stopBle" -> { ble?.stop(); ble = null; result.success(null) }
                "startSpp" -> {
                    spp?.stop()
                    spp = SppServer(rx("spp"), conn("spp")).also { it.start() }
                    result.success(null)
                }
                "stopSpp" -> { spp?.stop(); spp = null; result.success(null) }
                "send" -> {
                    val t = call.argument<String>("transport")
                    val bytes = call.argument<ByteArray>("bytes") ?: ByteArray(0)
                    if (t == "ble") ble?.send(bytes) else spp?.send(bytes)
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        }
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(args: Any?, sink: EventChannel.EventSink) { events = sink }
            override fun onCancel(args: Any?) { events = null }
        })
    }

    // 注: rx は transport を束縛した関数を返すヘルパ
    private fun rx(transport: String): (ByteArray) -> Unit = { bytes ->
        main.post {
            events?.success(mapOf("type" to "rx", "transport" to transport, "bytes" to bytes.toList()))
        }
    }

    private fun conn(transport: String): (String, String) -> Unit = { state, device ->
        main.post {
            events?.success(mapOf("type" to "conn", "transport" to transport, "state" to state, "device" to device))
        }
    }
}

File: macos/Runner/Elm327Plugin.swift
import FlutterMacOS
import Foundation

class Elm327Plugin: NSObject, FlutterStreamHandler {
    private var sink: FlutterEventSink?
    private var ble: BleGattServer?

    func register(messenger: FlutterBinaryMessenger) {
        let control = FlutterMethodChannel(name: "elm327/control", binaryMessenger: messenger)
        let events = FlutterEventChannel(name: "elm327/events", binaryMessenger: messenger)
        events.setStreamHandler(self)
        control.setMethodCallHandler { [weak self] call, result in
            guard let self = self else { return }
            switch call.method {
            case "capabilities":
                result(["ble"])
            case "startBle":
                let args = call.arguments as? [String: Any]
                let fff0 = (args?["profile"] as? String) == "fff0"
                self.ble?.stop()
                self.ble = BleGattServer(
                    onRx: { data in self.emitRx(Array(data)) },
                    onConn: { s, d in self.emitConn(s, d) })
                self.ble?.start(useFff0: fff0)
                result(nil)
            case "stopBle":
                self.ble?.stop()
                self.ble = nil
                result(nil)
            case "startSpp", "stopSpp":
                result(FlutterError(code: "unsupported",
                                    message: "SPP not supported on macOS",
                                    details: nil))
            case "send":
                let args = call.arguments as? [String: Any]
                if let data = args?["bytes"] as? FlutterStandardTypedData {
                    self.ble?.send(data.data)
                }
                result(nil)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }

    private func emitRx(_ bytes: [UInt8]) {
        DispatchQueue.main.async {
            self.sink?(["type": "rx", "transport": "ble", "bytes": bytes.map { Int($0) }])
        }
    }

    private func emitConn(_ state: String, _ device: String) {
        DispatchQueue.main.async {
            self.sink?(["type": "conn", "transport": "ble", "state": state, "device": device])
        }
    }

    func onListen(withArguments _: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        return nil
    }

    func onCancel(withArguments _: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
